### PR TITLE
[SYCL] sycl::kernel::get_kernel_bundle() may return a kernel_bundle with a null impl.

### PR DESCRIPTION
**Pull request by zahiraam**
This is an issue complaining about that return from getSyclObjImpl can be null.
I talked to @steffenlarsen off line and although this is considered to be impossible (every constructor should create for itself and impl), it doesn't seem to be enforced anywhere in the code. We decided that would be a good compromise to make sure that impl is always non-null.

### CODE REVIEW

**Review by againull — DISMISSED** [review]

**Comment by againull** [comment]
@zahiraam It looks like this change causes multiple LIT test failures, please address them.

File: sycl/source/handler.cpp [hunk 1]
```diff
@@ -77,11 +77,25 @@ handler::getOrInsertHandlerKernelBundle(bool Insert) const {
 }
 
 // Sets kernel bundle to the provided one.
+// If Kernel doesn't a kernel_bundle associated with it, the handler would
```
Comment (aelovikov-intel) [inline sycl/source/handler.cpp:80]: "doesn't have" ?

File: sycl/source/handler.cpp [hunk 2]
```diff
@@ -77,11 +77,25 @@ handler::getOrInsertHandlerKernelBundle(bool Insert) const {
...
+// still create a kernel_bundlxbe from a nullptr kernel_bundle_impl. This would
```
Comment (aelovikov-intel) [inline sycl/source/handler.cpp:81]:
```suggestion
// still create a kernel_bundle from a nullptr kernel_bundle_impl. This would
```

File: sycl/source/handler.cpp [hunk 3]
```diff
@@ -77,11 +77,25 @@ handler::getOrInsertHandlerKernelBundle(bool Insert) const {
...
+// have an invlaid impl and triggers an assertion in the handler. To remedy the
```
Comment (aelovikov-intel) [inline sycl/source/handler.cpp:82]:
```suggestion
// have an invalid impl and trigger an assertion in the handler. To remedy the
```

**Review by aelovikov-intel — COMMENTED** [review]
I wonder if that is just a warning or a real issue it found. If the latter, can we update the PR's title?

**Comment by zahiraam** [comment]
@aelovikov-intel Is the LIT fail a know issue? I don't see that's related to the patch?

**Comment by aelovikov-intel** [comment]
I think it's related to this patch - you've added a new method that is visible outside libsycl.so. The failing test has instructions on how to update the CHECKs.

**Comment by zahiraam** [comment]
> I think it's related to this patch - you've added a new method that is visible outside libsycl.so. The failing test has instructions on how to update the CHECKs.

This is the command that I am using to add the symbol, but I am getting an error:
bash-4.4$ python3 sycl/tools/abi_check.py --mode dump_symbols --output ./output.dump /export/iusers/zahiraam/my-sycl/debugbuild/lib/libsycl.so
Traceback (most recent call last):
  File "sycl/tools/abi_check.py", line 156, in <module>
    main()
  File "sycl/tools/abi_check.py", line 152, in main
    dump_symbols(args.target_library, args.output)
  File "sycl/tools/abi_check.py", line 78, in dump_symbols
    readobj_opts, target_path])
  File "/nfs/sc/proj/icl/rdrive/ics/itools/pkgtools/python/v3_7_8/efi2_rhel70/lib/python3.7/subprocess.py", line 411, in check_output
    **kwargs).stdout
  File "/nfs/sc/proj/icl/rdrive/ics/itools/pkgtools/python/v3_7_8/efi2_rhel70/lib/python3.7/subprocess.py", line 488, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/nfs/sc/proj/icl/rdrive/ics/itools/pkgtools/python/v3_7_8/efi2_rhel70/lib/python3.7/subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "/nfs/sc/proj/icl/rdrive/ics/itools/pkgtools/python/v3_7_8/efi2_rhel70/lib/python3.7/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'llvm-readobj': 'llvm-readobj'
bash-4.4$
Is there an env variable that needs to be set to run the command?
Thanks.

**Comment by steffenlarsen** [comment]
CUDA failure in Assert/assert_in_simultaneously_multiple_tus.cpp is known and the test has since been disabled.

**Comment by zahiraam** [comment]
> CUDA failure in Assert/assert_in_simultaneously_multiple_tus.cpp is known and the test has since been disabled.

Thanks @steffenlarsen . @againull Can this be checked in? Thanks.

File: sycl/source/handler.cpp [hunk 4]
```diff
@@ -77,11 +77,25 @@ handler::getOrInsertHandlerKernelBundle(bool Insert) const {
 }
 
 // Sets kernel bundle to the provided one.
+// If Kernel doesn't have a kernel_bundle associated with it, the handler would
+// still create a kernel_bundle from a nullptr kernel_bundle_impl. This would
+// have an invalid impl and trigger an assertion in the handler. To remedy the
+// issue, we added the (overloaded) function below that gets the
+// kernel_bundle_impl associated with the kernel and then calls the other
+// version setHandlerKernelBundle with that.
```
Comment (aelovikov-intel) [inline sycl/source/handler.cpp:85]: I don't think this is relevant for this particular overload and would suggest to drop it.

File: sycl/source/handler.cpp [hunk 5]
```diff
@@ -77,11 +77,25 @@ handler::getOrInsertHandlerKernelBundle(bool Insert) const {
...
 void handler::setHandlerKernelBundle(
     const std::shared_ptr<detail::kernel_bundle_impl> &NewKernelBundleImpPtr) {
   MImpl->MKernelBundle = NewKernelBundleImpPtr;
 }
 
+void handler::setHandlerKernelBundle(kernel Kernel) {
+  // Kernel may not have an associated kernel bundle if it is created from a
+  // program.
```
Comment (aelovikov-intel) [inline sycl/source/handler.cpp:93]:
```suggestion
  // program. As such, apply getSyclObjImpl directly on the kernel, i.e. not
  //  the other way around: getSyclObjImp(Kernel->get_kernel_bundle()).
```

**Review by aelovikov-intel — COMMENTED** [review]
Functionally this change is good. I'd change some comments (see inline) but it might be just me.

@againull , WDYT?

**Comment by againull** [comment]
> Functionally this change is good. I'd change some comments (see inline) but it might be just me.
> 
> @againull , WDYT?

I agree  that it would be nice to change the comments per suggestions. @zahiraam Could you please update comments per suggestions and I will merge afterwards.

**Comment by zahiraam** [comment]
> > Functionally this change is good. I'd change some comments (see inline) but it might be just me.
> > @againull , WDYT?
> 
> I agree that it would be nice to change the comments per suggestions. @zahiraam Could you please update comments per suggestions and I will merge afterwards.

Thanks @aelovikov-intel  and @againull. Done.

**Review by aelovikov-intel — DISMISSED** [review]
Thanks!

**Comment by gmlueck** [comment]
This PR seems to have some unrelated changes, for example those changes in "sycl/doc/design/OptionalDeviceFeatures.md".  Is this due to a merge mistake?

**Comment by steffenlarsen** [comment]
> This PR seems to have some unrelated changes, for example those changes in "sycl/doc/design/OptionalDeviceFeatures.md". Is this due to a merge mistake?

Apologies, @gmlueck. There was some problems in a rebase. We are fixing it now.

**Comment by zahiraam** [comment]
The PR is now clean. Thanks to @steffenlarsen who spent time looking at my ws and cleaning it up!

**Review by steffenlarsen — APPROVED** [review]
Apart from the rebase and squash, there are no changes since previous approvals.

**Comment by zahiraam** [comment]
> This PR seems to have some unrelated changes, for example those changes in "sycl/doc/design/OptionalDeviceFeatures.md". Is this due to a merge mistake?

Yes it is. Sorry about the confusion.

File: sycl/source/handler.cpp [hunk 6]
```diff
@@ -82,6 +82,15 @@ void handler::setHandlerKernelBundle(
   MImpl->MKernelBundle = NewKernelBundleImpPtr;
 }
 
+void handler::setHandlerKernelBundle(kernel Kernel) {
+  // Kernel may not have an associated kernel bundle if it is created from a
+  // program. As such, apply getSyclObjImpl directly on the kernel, i.e. not
+  //  the other way around: getSyclObjImp(Kernel->get_kernel_bundle()).
```
Comment (gmlueck) [inline sycl/source/handler.cpp:88]: I think this is not directly related to your change, but this comment makes me think there is a problem someplace.  Why is there no kernel bundle for a kernel created from a program?  The SYCL spec defines `kernel::get_kernel_bundle()` for all kernels, even those created from backend interop.

Comment (steffenlarsen) [reply sycl/source/handler.cpp:88]: Arguably there is, but since program is being removed I don't think it is worth the worry.

Comment (gmlueck) [reply sycl/source/handler.cpp:88]: OK, I was wondering if the issue about having no kernel bundle is limited to just the deprecated `program` case, or if it also included other cases like constructing a kernel from backend interop.  If it is only the `program` case, I agree there is no need to fix it now.  We should instead wait until `program` is removed.

Comment (steffenlarsen) [reply sycl/source/handler.cpp:88]: This problem should be unique to kernel bundles in kernels created from `sycl::program`. For `sycl::make_kernel` we either use a supplied kernel bundle (part of the L0 input type) or an empty "interop" kernel bundle.
